unit tests fail

The unit tests fail on the actual branch in internal/pkg/node/util_test.go

Here is some more info. This looks related to the data structure changes.
- This is an access violation.
- n.Kernel is nil here: https://github.com/hpcng/warewulf/blob/95050a44cebeb5fe90cb6ef3a2c3f7e24543b812/internal/pkg/node/constructors.go#L70
- The failing test is using this configuration. Kernel does not appear to be defined: https://github.com/hpcng/warewulf/blob/95050a44cebeb5fe90cb6ef3a2c3f7e24543b812/internal/pkg/node/util_test.go#L10
```
[mhink@rocky warewulf]$ git log -n 1
commit 6c0fd068cc7118b8bf9a3f9e901b179a30474858 (HEAD -> main, origin/main, origin/HEAD)
Author: Christian Goll <cgoll@suse.de>
Date:   Thu Apr 7 17:22:38 2022 +0200

    updated Makefile for v4.3.0-rc1
[mhink@rocky warewulf]$ make test-it
go test -v ./...
?   	github.com/hpcng/warewulf/cmd/bash_completion	[no test files]
?   	github.com/hpcng/warewulf/cmd/config_defaults	[no test files]
?   	github.com/hpcng/warewulf/cmd/man_page	[no test files]
?   	github.com/hpcng/warewulf/cmd/wwclient	[no test files]
?   	github.com/hpcng/warewulf/cmd/wwctl	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwclient	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/configure	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/configure/dhcp	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/configure/nfs	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/configure/ssh	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/configure/tftp	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container	[no test files]
=== RUN   TestArgsContainerBuild
--- PASS: TestArgsContainerBuild (0.00s)
PASS
ok  	github.com/hpcng/warewulf/internal/app/wwctl/container/build	(cached)
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/delete	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/exec	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/exec/child	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/imprt	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/list	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/shell	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/show	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/container/syncuser	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/kernel	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/kernel/delete	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/kernel/imprt	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/kernel/list	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/add	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/console	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/delete	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/list	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/sensors	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/set	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/node/status	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/build	[no test files]
=== RUN   TestArgsOverlayChmod
--- PASS: TestArgsOverlayChmod (0.00s)
PASS
ok  	github.com/hpcng/warewulf/internal/app/wwctl/overlay/chmod	(cached)
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/chown	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/create	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/delete	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/edit	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/imprt	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/list	[no test files]
=== RUN   TestArgsOverlayMkdir
--- PASS: TestArgsOverlayMkdir (0.00s)
PASS
ok  	github.com/hpcng/warewulf/internal/app/wwctl/overlay/mkdir	(cached)
?   	github.com/hpcng/warewulf/internal/app/wwctl/overlay/show	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/cycle	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/off	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/on	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/reset	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/soft	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/power/status	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/profile	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/profile/add	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/profile/delete	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/profile/list	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/profile/set	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server/reload	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server/restart	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server/start	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server/status	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/server/stop	[no test files]
?   	github.com/hpcng/warewulf/internal/app/wwctl/version	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/batch	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/buildconfig	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/configure	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/container	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/help	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/kernel	[no test files]
=== RUN   Test_nodeYaml_FindByHwaddr
=== RUN   Test_nodeYaml_FindByHwaddr/emptyString
=== RUN   Test_nodeYaml_FindByHwaddr/noIpString
=== RUN   Test_nodeYaml_FindByHwaddr/intString
=== RUN   Test_nodeYaml_FindByHwaddr/invalidMAC
=== RUN   Test_nodeYaml_FindByHwaddr/validMACNotFound
--- FAIL: Test_nodeYaml_FindByHwaddr (0.00s)
    --- PASS: Test_nodeYaml_FindByHwaddr/emptyString (0.00s)
    --- PASS: Test_nodeYaml_FindByHwaddr/noIpString (0.00s)
    --- PASS: Test_nodeYaml_FindByHwaddr/intString (0.00s)
    --- PASS: Test_nodeYaml_FindByHwaddr/invalidMAC (0.00s)
    --- FAIL: Test_nodeYaml_FindByHwaddr/validMACNotFound (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x56c1aa]

goroutine 11 [running]:
testing.tRunner.func1.2({0x58fde0, 0x6f0230})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x58fde0, 0x6f0230})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/hpcng/warewulf/internal/pkg/node.(*nodeYaml).FindAllNodes(0xc0000e7648)
	/home/mhink/go/src/github.com/hpcng/warewulf/internal/pkg/node/constructors.go:70 +0x36a
github.com/hpcng/warewulf/internal/pkg/node.(*nodeYaml).FindByHwaddr(_, {_, _})
	/home/mhink/go/src/github.com/hpcng/warewulf/internal/pkg/node/util.go:16 +0x85
github.com/hpcng/warewulf/internal/pkg/node.Test_nodeYaml_FindByHwaddr.func1(0xc00008f860)
	/home/mhink/go/src/github.com/hpcng/warewulf/internal/pkg/node/util_test.go:88 +0x8b
testing.tRunner(0xc00008f860, 0xc0000517c0)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
FAIL	github.com/hpcng/warewulf/internal/pkg/node	0.005s
?   	github.com/hpcng/warewulf/internal/pkg/oci	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/overlay	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/pidfile	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/power	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/util	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/version	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/warewulfconf	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/warewulfd	[no test files]
?   	github.com/hpcng/warewulf/internal/pkg/wwlog	[no test files]
?   	github.com/hpcng/warewulf/pkg/hostlist	[no test files]
?   	github.com/hpcng/warewulf/test	[no test files]
FAIL
make: *** [Makefile:145: test-it] Error 1
[mhink@rocky warewulf]$ 
```

--- a/internal/pkg/node/constructors.go
+++ b/internal/pkg/node/constructors.go
@@ -297,25 +297,28 @@ func (config *nodeYaml) FindAllProfiles() ([]NodeInfo, error) {
 		p.ContainerName.Set(profile.ContainerName)
 		p.Ipxe.Set(profile.Ipxe)
 		p.Init.Set(profile.Init)
-		p.Kernel.Args.Set(profile.Kernel.Args)
-		p.Ipmi.Netmask.Set(profile.Ipmi.Netmask)
-		p.Ipmi.Port.Set(profile.Ipmi.Port)
-		p.Ipmi.Gateway.Set(profile.Ipmi.Gateway)
-		p.Ipmi.UserName.Set(profile.Ipmi.UserName)
-		p.Ipmi.Password.Set(profile.Ipmi.Password)
-		p.Ipmi.Interface.Set(profile.Ipmi.Interface)
-		p.Ipmi.Write.SetB(profile.Ipmi.Write)
+		if profile.Kernel != nil {
+			p.Kernel.Args.Set(profile.Kernel.Args)
+			if profile.Kernel.Override != "" {
+				p.Kernel.Override.Set(profile.Kernel.Override)
+			} else if profile.Kernel.Version != "" {
+				p.Kernel.Override.Set(profile.Kernel.Version)
+			}
+		}
+		if profile.Ipmi != nil {
+			p.Ipmi.Netmask.Set(profile.Ipmi.Netmask)
+			p.Ipmi.Port.Set(profile.Ipmi.Port)
+			p.Ipmi.Gateway.Set(profile.Ipmi.Gateway)
+			p.Ipmi.UserName.Set(profile.Ipmi.UserName)
+			p.Ipmi.Password.Set(profile.Ipmi.Password)
+			p.Ipmi.Interface.Set(profile.Ipmi.Interface)
+			p.Ipmi.Write.SetB(profile.Ipmi.Write)
+		}
 		p.RuntimeOverlay.SetSlice(profile.RuntimeOverlay)
 		p.SystemOverlay.SetSlice(profile.SystemOverlay)
 		p.Root.Set(profile.Root)
 		p.AssetKey.Set(profile.AssetKey)
 		p.Discoverable.Set(profile.Discoverable)
-
-		if profile.Kernel.Override != "" {
-			p.Kernel.Override.Set(profile.Kernel.Override)
-		} else if profile.Kernel.Version != "" {
-			p.Kernel.Override.Set(profile.Kernel.Version)
-		}
 
 		for devname, netdev := range profile.NetDevs {
 			if _, ok := p.NetDevs[devname]; !ok {

--- a/internal/pkg/node/constructors.go
+++ b/internal/pkg/node/constructors.go
@@ -62,6 +62,8 @@ func (config *nodeYaml) FindAllNodes() ([]NodeInfo, error) {
 		wwlog.Printf(wwlog.DEBUG, "In node loop: %s\n", nodename)
 		n.NetDevs = make(map[string]*NetDevEntry)
 		n.Tags = make(map[string]*Entry)
+		n.Kernel = new(KernelEntry)
+		n.Ipmi = new(IpmiEntry)
 		n.SystemOverlay.SetDefault("wwinit")
 		n.RuntimeOverlay.SetDefault("generic")
 		n.Ipxe.SetDefault("default")
@@ -110,7 +112,7 @@ func (config *nodeYaml) FindAllNodes() ([]NodeInfo, error) {
 		n.Discoverable.Set(node.Discoverable)
 
 		if node.Kernel != nil {
-		if node.Kernel.Override != "" {
+			if node.Kernel.Override != "" {
 				n.Kernel.Override.Set(node.Kernel.Override)
 			} else if node.Kernel.Version != "" {
 				n.Kernel.Override.Set(node.Kernel.Version)

--- a/internal/pkg/node/constructors.go
+++ b/internal/pkg/node/constructors.go
@@ -289,7 +289,8 @@ func (config *nodeYaml) FindAllProfiles() ([]NodeInfo, error) {
 		var p NodeInfo
 		p.NetDevs = make(map[string]*NetDevEntry)
 		p.Tags = make(map[string]*Entry)
-
+		p.Kernel = new(KernelEntry)
+		p.Ipmi = new(IpmiEntry)
 		p.Id.Set(name)
 		p.Comment.Set(profile.Comment)
 		p.ClusterName.Set(profile.ClusterName)

--- a/internal/pkg/node/constructors.go
+++ b/internal/pkg/node/constructors.go
@@ -67,6 +67,10 @@ func (config *nodeYaml) FindAllNodes() ([]NodeInfo, error) {
 		n.Ipxe.SetDefault("default")
 		n.Init.SetDefault("/sbin/init")
 		n.Root.SetDefault("initramfs")
+
+		if n.Kernel == nil {
+			n.Kernel = &KernelEntry{}
+		}
 		n.Kernel.Args.SetDefault("quiet crashkernel=no vga=791")
 
 		fullname := strings.SplitN(nodename, ".", 2)
@@ -83,28 +87,34 @@ func (config *nodeYaml) FindAllNodes() ([]NodeInfo, error) {
 		n.Id.Set(nodename)
 		n.Comment.Set(node.Comment)
 		n.ContainerName.Set(node.ContainerName)
-		n.Kernel.Args.Set(node.Kernel.Args)
+		if node.Kernel != nil {
+			n.Kernel.Args.Set(node.Kernel.Args)
+		}
 		n.ClusterName.Set(node.ClusterName)
 		n.Ipxe.Set(node.Ipxe)
 		n.Init.Set(node.Init)
-		n.Ipmi.Ipaddr.Set(node.Ipmi.Ipaddr)
-		n.Ipmi.Netmask.Set(node.Ipmi.Netmask)
-		n.Ipmi.Port.Set(node.Ipmi.Port)
-		n.Ipmi.Gateway.Set(node.Ipmi.Gateway)
-		n.Ipmi.UserName.Set(node.Ipmi.UserName)
-		n.Ipmi.Password.Set(node.Ipmi.Password)
-		n.Ipmi.Interface.Set(node.Ipmi.Interface)
-		n.Ipmi.Write.SetB(node.Ipmi.Write)
+		if node.Ipmi != nil {
+			n.Ipmi.Ipaddr.Set(node.Ipmi.Ipaddr)
+			n.Ipmi.Netmask.Set(node.Ipmi.Netmask)
+			n.Ipmi.Port.Set(node.Ipmi.Port)
+			n.Ipmi.Gateway.Set(node.Ipmi.Gateway)
+			n.Ipmi.UserName.Set(node.Ipmi.UserName)
+			n.Ipmi.Password.Set(node.Ipmi.Password)
+			n.Ipmi.Interface.Set(node.Ipmi.Interface)
+			n.Ipmi.Write.SetB(node.Ipmi.Write)
+		}
 		n.SystemOverlay.SetSlice(node.SystemOverlay)
 		n.RuntimeOverlay.SetSlice(node.RuntimeOverlay)
 		n.Root.Set(node.Root)
 		n.AssetKey.Set(node.AssetKey)
 		n.Discoverable.Set(node.Discoverable)
 
+		if node.Kernel != nil {
 		if node.Kernel.Override != "" {
-			n.Kernel.Override.Set(node.Kernel.Override)
-		} else if node.Kernel.Version != "" {
-			n.Kernel.Override.Set(node.Kernel.Version)
+				n.Kernel.Override.Set(node.Kernel.Override)
+			} else if node.Kernel.Version != "" {
+				n.Kernel.Override.Set(node.Kernel.Version)
+			}
 		}
 
 		for devname, netdev := range node.NetDevs {
@@ -179,27 +189,33 @@ func (config *nodeYaml) FindAllNodes() ([]NodeInfo, error) {
 			n.Comment.SetAlt(config.NodeProfiles[p].Comment, p)
 			n.ClusterName.SetAlt(config.NodeProfiles[p].ClusterName, p)
 			n.ContainerName.SetAlt(config.NodeProfiles[p].ContainerName, p)
-			n.Kernel.Args.SetAlt(config.NodeProfiles[p].Kernel.Args, p)
+			if config.NodeProfiles[p].Kernel != nil {
+				n.Kernel.Args.SetAlt(config.NodeProfiles[p].Kernel.Args, p)
+			}
 			n.Ipxe.SetAlt(config.NodeProfiles[p].Ipxe, p)
 			n.Init.SetAlt(config.NodeProfiles[p].Init, p)
-			n.Ipmi.Ipaddr.SetAlt(config.NodeProfiles[p].Ipmi.Ipaddr, p)
-			n.Ipmi.Netmask.SetAlt(config.NodeProfiles[p].Ipmi.Netmask, p)
-			n.Ipmi.Port.SetAlt(config.NodeProfiles[p].Ipmi.Port, p)
-			n.Ipmi.Gateway.SetAlt(config.NodeProfiles[p].Ipmi.Gateway, p)
-			n.Ipmi.UserName.SetAlt(config.NodeProfiles[p].Ipmi.UserName, p)
-			n.Ipmi.Password.SetAlt(config.NodeProfiles[p].Ipmi.Password, p)
-			n.Ipmi.Interface.SetAlt(config.NodeProfiles[p].Ipmi.Interface, p)
-			n.Ipmi.Write.SetB(config.NodeProfiles[p].Ipmi.Write)
+			if config.NodeProfiles[p].Ipmi != nil {
+				n.Ipmi.Ipaddr.SetAlt(config.NodeProfiles[p].Ipmi.Ipaddr, p)
+				n.Ipmi.Netmask.SetAlt(config.NodeProfiles[p].Ipmi.Netmask, p)
+				n.Ipmi.Port.SetAlt(config.NodeProfiles[p].Ipmi.Port, p)
+				n.Ipmi.Gateway.SetAlt(config.NodeProfiles[p].Ipmi.Gateway, p)
+				n.Ipmi.UserName.SetAlt(config.NodeProfiles[p].Ipmi.UserName, p)
+				n.Ipmi.Password.SetAlt(config.NodeProfiles[p].Ipmi.Password, p)
+				n.Ipmi.Interface.SetAlt(config.NodeProfiles[p].Ipmi.Interface, p)
+				n.Ipmi.Write.SetB(config.NodeProfiles[p].Ipmi.Write)
+			}
 			n.SystemOverlay.SetAltSlice(config.NodeProfiles[p].SystemOverlay, p)
 			n.RuntimeOverlay.SetAltSlice(config.NodeProfiles[p].RuntimeOverlay, p)
 			n.Root.SetAlt(config.NodeProfiles[p].Root, p)
 			n.AssetKey.SetAlt(config.NodeProfiles[p].AssetKey, p)
 			n.Discoverable.SetAlt(config.NodeProfiles[p].Discoverable, p)
 
-			if config.NodeProfiles[p].Kernel.Override != "" {
-				n.Kernel.Override.SetAlt(config.NodeProfiles[p].Kernel.Override, p)
-			} else if config.NodeProfiles[p].Kernel.Version != "" {
-				n.Kernel.Override.SetAlt(config.NodeProfiles[p].Kernel.Version, p)
+			if config.NodeProfiles[p].Kernel != nil {
+				if config.NodeProfiles[p].Kernel.Override != "" {
+					n.Kernel.Override.SetAlt(config.NodeProfiles[p].Kernel.Override, p)
+				} else if config.NodeProfiles[p].Kernel.Version != "" {
+					n.Kernel.Override.SetAlt(config.NodeProfiles[p].Kernel.Version, p)
+				}
 			}
 
 			for devname, netdev := range config.NodeProfiles[p].NetDevs {

--- a/internal/pkg/node/methods.go
+++ b/internal/pkg/node/methods.go
@@ -209,6 +209,16 @@ func (ent *Entry) GetRealSlice() []string {
 	return ent.value
 }
 
+/*
+true if the entry has set a real value, else false.
+*/
+func (ent *Entry) GotReal() bool {
+	if len(ent.value) == 0 {
+		return false
+	}
+	return true
+}
+
 /**********
  *
  * Misc

--- a/internal/pkg/node/methods.go
+++ b/internal/pkg/node/methods.go
@@ -213,10 +213,7 @@ func (ent *Entry) GetRealSlice() []string {
 true if the entry has set a real value, else false.
 */
 func (ent *Entry) GotReal() bool {
-	if len(ent.value) == 0 {
-		return false
-	}
-	return true
+	return len(ent.value) != 0
 }
 
 /**********

--- a/internal/pkg/node/modifiers.go
+++ b/internal/pkg/node/modifiers.go
@@ -28,10 +28,11 @@ func (config *nodeYaml) AddNode(nodeID string) (NodeInfo, error) {
 	config.Nodes[nodeID] = &node
 	config.Nodes[nodeID].Profiles = []string{"default"}
 	config.Nodes[nodeID].NetDevs = make(map[string]*NetDevs)
-
 	n.Id.Set(nodeID)
 	n.Profiles = []string{"default"}
 	n.NetDevs = make(map[string]*NetDevEntry)
+	n.Ipmi = new(IpmiEntry)
+	n.Kernel = new(KernelEntry)
 
 	return n, nil
 }
@@ -60,16 +61,24 @@ func (config *nodeYaml) NodeUpdate(node NodeInfo) error {
 	config.Nodes[nodeID].ClusterName = node.ClusterName.GetReal()
 	config.Nodes[nodeID].Ipxe = node.Ipxe.GetReal()
 	config.Nodes[nodeID].Init = node.Init.GetReal()
-	config.Nodes[nodeID].Kernel.Override = node.Kernel.Override.GetReal()
-	config.Nodes[nodeID].Kernel.Args = node.Kernel.Args.GetReal()
-	config.Nodes[nodeID].Ipmi.Ipaddr = node.Ipmi.Ipaddr.GetReal()
-	config.Nodes[nodeID].Ipmi.Netmask = node.Ipmi.Netmask.GetReal()
-	config.Nodes[nodeID].Ipmi.Port = node.Ipmi.Port.GetReal()
-	config.Nodes[nodeID].Ipmi.Gateway = node.Ipmi.Gateway.GetReal()
-	config.Nodes[nodeID].Ipmi.UserName = node.Ipmi.UserName.GetReal()
-	config.Nodes[nodeID].Ipmi.Password = node.Ipmi.Password.GetReal()
-	config.Nodes[nodeID].Ipmi.Interface = node.Ipmi.Interface.GetReal()
-	config.Nodes[nodeID].Ipmi.Write = node.Ipmi.Write.GetB()
+	if node.Kernel.Override.GotReal() || node.Kernel.Args.GotReal() {
+		config.Nodes[nodeID].Kernel = new(KernelConf)
+		config.Nodes[nodeID].Kernel.Override = node.Kernel.Override.GetReal()
+		config.Nodes[nodeID].Kernel.Args = node.Kernel.Args.GetReal()
+	}
+	if node.Ipmi.Ipaddr.GotReal() || node.Ipmi.Netmask.GotReal() ||
+		node.Ipmi.Port.GotReal() || node.Ipmi.Gateway.GotReal() || node.Ipmi.UserName.GotReal() ||
+		node.Ipmi.Password.GotReal() || node.Ipmi.Interface.GotReal() || node.Ipmi.Write.GotReal() {
+		config.Nodes[nodeID].Ipmi = new(IpmiConf)
+		config.Nodes[nodeID].Ipmi.Ipaddr = node.Ipmi.Ipaddr.GetReal()
+		config.Nodes[nodeID].Ipmi.Netmask = node.Ipmi.Netmask.GetReal()
+		config.Nodes[nodeID].Ipmi.Port = node.Ipmi.Port.GetReal()
+		config.Nodes[nodeID].Ipmi.Gateway = node.Ipmi.Gateway.GetReal()
+		config.Nodes[nodeID].Ipmi.UserName = node.Ipmi.UserName.GetReal()
+		config.Nodes[nodeID].Ipmi.Password = node.Ipmi.Password.GetReal()
+		config.Nodes[nodeID].Ipmi.Interface = node.Ipmi.Interface.GetReal()
+		config.Nodes[nodeID].Ipmi.Write = node.Ipmi.Write.GetB()
+	}
 	config.Nodes[nodeID].RuntimeOverlay = node.RuntimeOverlay.GetRealSlice()
 	config.Nodes[nodeID].SystemOverlay = node.SystemOverlay.GetRealSlice()
 	config.Nodes[nodeID].Root = node.Root.GetReal()

